File: onion-architecture/java/src/test/java/com/example/infrastructure/adapter/StudentRepositoryImplTest.java
package com.example.infrastructure.adapter;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotSame;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.example.domain.Course;
import com.example.domain.Student;

class StudentRepositoryImplTest {

    private StudentRepositoryImpl repository;
    private Student student;

    @BeforeEach
    void setUp() {
        repository = new StudentRepositoryImpl();
        student = new Student(1L, "John Doe");
    }

    @Test
    void shouldReturnEmptyOptionalForNonExistentStudent() {
        Optional<Student> result = repository.findById(1L);

        assertFalse(result.isPresent());
    }

    @Test
    void shouldSaveAndFindStudentById() {
        repository.save(student);

        Optional<Student> result = repository.findById(1L);

        assertTrue(result.isPresent());
        assertEquals(student.getId(), result.get().getId());
        assertSame(student, result.get());
    }

    @Test
    void shouldUpdateExistingStudent() {
        repository.save(student);

        Course course = new Course(1L, "Java Programming");
        student.enrol(course);
        repository.save(student);

        Optional<Student> result = repository.findById(1L);

        assertTrue(result.isPresent());
        assertTrue(result.get().getEnrolledCourses().contains(course));
        assertEquals(1, result.get().getEnrolledCourses().size());
    }

    @Test
    void shouldSaveMultipleStudents() {
        Student student2 = new Student(2L, "Jane Doe");

        repository.save(student);
        repository.save(student2);

        Optional<Student> result1 = repository.findById(1L);
        Optional<Student> result2 = repository.findById(2L);

        assertTrue(result1.isPresent());
        assertTrue(result2.isPresent());
        assertEquals(1L, result1.get().getId());
        assertEquals(2L, result2.get().getId());
    }

    @Test
    void shouldReturnEmptyOptionalForNullId() {
        Optional<Student> result = repository.findById(null);

        assertFalse(result.isPresent());
    }

    @Test
    void shouldOverwriteStudentWhenSavingWithSameId() {
        repository.save(student);

        Student newStudent = new Student(1L, "John Smith");
        repository.save(newStudent);

        Optional<Student> result = repository.findById(1L);

        assertTrue(result.isPresent());
        assertSame(newStudent, result.get());
        assertNotSame(student, result.get());
    }

    @Test
    void shouldPersistStudentEnrolments() {
        Course javaCourse = new Course(1L, "Java Programming");
        Course springCourse = new Course(2L, "Spring Boot");

        student.enrol(javaCourse);
        student.enrol(springCourse);
        repository.save(student);

        Optional<Student> result = repository.findById(1L);

        assertTrue(result.isPresent());
        assertEquals(2, result.get().getEnrolledCourses().size());
        assertTrue(result.get().getEnrolledCourses().contains(javaCourse));
        assertTrue(result.get().getEnrolledCourses().contains(springCourse));
    }

    @Test
    void shouldHandleNullStudentSaveGracefully() {
        assertThrows(NullPointerException.class, () -> repository.save(null));
    }
}

File: onion-architecture/kotlin/src/test/kotlin/com/example/infrastructure/adapter/StudentRepositoryImplTest.kt
package com.example.infrastructure.adapter

import com.example.domain.Course
import com.example.domain.Student
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Assertions.*

class StudentRepositoryImplTest {

    private lateinit var repository: StudentRepositoryImpl
    private lateinit var student: Student

    @BeforeEach
    fun setUp() {
        repository = StudentRepositoryImpl()
        student = Student(1L, "John Doe")
    }

    @Test
    fun shouldReturnNullForNonExistentStudent() {
        val result = repository.findById(1L)

        assertNull(result)
    }

    @Test
    fun shouldSaveAndFindStudentById() {
        repository.save(student)

        val result = repository.findById(1L)

        assertNotNull(result)
        assertEquals(student.id, result!!.id)
        assertSame(student, result)
    }

    @Test
    fun shouldUpdateExistingStudent() {
        repository.save(student)

        val course = Course(1L, "Java Programming")
        student.enrol(course)
        repository.save(student)

        val result = repository.findById(1L)

        assertNotNull(result)
        assertTrue(result!!.getEnrolledCourses().contains(course))
        assertEquals(1, result.getEnrolledCourses().size)
    }

    @Test
    fun shouldSaveMultipleStudents() {
        val student2 = Student(2L, "Jane Doe")

        repository.save(student)
        repository.save(student2)

        val result1 = repository.findById(1L)
        val result2 = repository.findById(2L)

        assertNotNull(result1)
        assertNotNull(result2)
        assertEquals(1L, result1!!.id)
        assertEquals(2L, result2!!.id)
    }

    @Test
    fun shouldReturnNullForZeroId() {
        val result = repository.findById(0L)

        assertNull(result)
    }

    @Test
    fun shouldOverwriteStudentWhenSavingWithSameId() {
        repository.save(student)

        val newStudent = Student(1L, "John Smith")
        repository.save(newStudent)

        val result = repository.findById(1L)

        assertNotNull(result)
        assertSame(newStudent, result)
        assertNotSame(student, result)
    }

    @Test
    fun shouldPersistStudentEnrolments() {
        val javaCourse = Course(1L, "Java Programming")
        val springCourse = Course(2L, "Spring Boot")

        student.enrol(javaCourse)
        student.enrol(springCourse)
        repository.save(student)

        val result = repository.findById(1L)

        assertNotNull(result)
        assertEquals(2, result!!.getEnrolledCourses().size)
        assertTrue(result.getEnrolledCourses().contains(javaCourse))
        assertTrue(result.getEnrolledCourses().contains(springCourse))
    }
}
